crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use sui_types::committee::EpochId;
+use typed_store::Map;
+
+use std::sync::Arc;
+use sui_types::base_types::ObjectDigest;
+
+use fastcrypto::hash::{Digest, MultisetHash};
+use sui_types::accumulator::Accumulator;
+use sui_types::error::SuiResult;
+use sui_types::messages::TransactionEffects;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use typed_store::rocks::TypedStoreError;
+
+use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use crate::authority::AuthorityStore;
+
+#[derive(Debug, Clone)]
+pub struct State {
+    pub effects: Vec<TransactionEffects>,
+    pub checkpoint_seq_num: CheckpointSequenceNumber,
+}
+
+pub struct StateAccumulator {
+    authority_store: Arc<AuthorityStore>,
+}
+
+impl StateAccumulator {
+    pub fn new(authority_store: Arc<AuthorityStore>) -> Self {
+        Self { authority_store }
+    }
+
+    pub fn accumulate_checkpoint(
+        &self,
+        state: State,
+        epoch_store: Arc<AuthorityPerEpochStore>,
+    ) -> SuiResult<Accumulator> {
+        if let Some(acc) = epoch_store.get_state_hash_for_checkpoint(&state.checkpoint_seq_num)? {
+            return Ok(acc);
+        }
+
+        let mut acc = Accumulator::default();
+
+        acc.insert_all(
+            state
+                .effects
+                .iter()
+                .flat_map(|fx| {
+                    fx.created
+                        .clone()
+                        .into_iter()
+                        .map(|(obj_ref, _)| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
+                })
+                .collect::<Vec<ObjectDigest>>(),
+        );
+        acc.remove_all(
+            state
+                .effects
+                .iter()
+                .flat_map(|fx| {
+                    fx.deleted
+                        .clone()
+                        .into_iter()
+                        .map(|obj_ref| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
+                })
+                .collect::<Vec<ObjectDigest>>(),
+        );
+        // TODO almost certainly not currectly handling "mutated" effects.
+        acc.insert_all(
+            state
+                .effects
+                .iter()
+                .flat_map(|fx| {
+                    fx.mutated
+                        .clone()
+                        .into_iter()
+                        .map(|(obj_ref, _)| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
+                })
+                .collect::<Vec<ObjectDigest>>(),
+        );
+
+        epoch_store.insert_state_hash_for_checkpoint(&state.checkpoint_seq_num, &acc)?;
+
+        epoch_store
```

Comment:
Note to self: realized there's a crash recovery issue here. If we crash after writing the accumulator but before `notify`, the next time this is called (on crash recovery), the func will early return and not notify, leaving a waiter waiting forever.

Make this transactional with writing the accumulator